Add rule to check the markup validity

qzhou1607 opened the pull request:
Fix #28

molant commented:
@qzhou1607 there are some conflicts. Can you please solve them?

molant commented on src/lib/rules/html-checker/html-checker.ts:
```diff
@@ -0,0 +1,107 @@
+        },
+        fixable: 'code',
+        recommended: true,
+        schema: [],
```
Doesn't the checker allow any configuration? I thought we could ignore some type of errors, etc.

molant commented on src/lib/rules/html-checker/html-checker.ts:
```diff
@@ -0,0 +1,107 @@
+            });
+
+            try {
+                await Promise.all(reportPromises);
```
Can we add support for async iterator? They should be working on typescript 2.3 (check the release notes, there might be some hack to do to make it work everywhere).

Also, I'd refactor the reporting and problem location code to another method and then call it in the iteration.

qzhou1607 commented:
@molant In this PR, I use the `html-validator` dependency, which uses the web service interface of the validator ( I doubled checked the web service and didn't see a limit of requests anywhere.). An alternative is to use the [vnu-jar](https://www.npmjs.com/package/vnu-jar), but using java with node seems complicated according to @alrra . Do you have any opinion on that which one to use?

sarvaje commented on src/lib/rule-context.ts:
```diff
@@ -107,7 +107,7 @@ export class RuleContext {
             this.id,
             severity || this.severity,
-            sourceCode,
+            sourceCode || codeSnippet,
```
IMO if the `codeSnippet` is present, maybe it should have preference over the `sourceCode`

sarvaje commented on src/lib/rules/html-checker/html-checker.ts:
```diff
@@ -0,0 +1,107 @@
+import { IRule, IRuleBuilder } from '../../types'; // eslint-disable-line no-unused-vars
+import { RuleContext } from '../../rule-context'; // eslint-disable-line no-unused-vars
+
+import { ITargetFetchEnd, IScanEnd, IProblemLocation, Severity } from '../../types'; // eslint-disable-line no-unused-vars
```
the `import` for `../../types` is duplicate, merge everything in just one `import`

sarvaje commented on src/lib/rules/html-checker/html-checker.ts:
```diff
@@ -0,0 +1,107 @@
+
+import { ITargetFetchEnd, IScanEnd, IProblemLocation, Severity } from '../../types'; // eslint-disable-line no-unused-vars
+
+const debug = d(__filename);
```
add the type to debug `debug.IDebugger`

austinkelleher commented on src/lib/rules/html-checker/html-checker.ts:
```diff
@@ -0,0 +1,107 @@
+            try {
+                result = await htmlCheckerPromise;
+            } catch (e) {
+                debug(`Error getting html checker result for ${resource}.`);
```
I think you should log the actual error.

austinkelleher replied:
Should be included in the report?

austinkelleher commented on src/lib/rules/html-checker/html-checker.ts:
```diff
@@ -0,0 +1,107 @@
+
+        const start = (data: ITargetFetchEnd) => {
+            const { response } = data;
+            const htmlChecker = require('html-validator');
```
Why is this `require` inside of the function instead of lifted to the top of the scope. It should also be an `import` right?

qzhou1607 replied:
@austinkelleher This is due to the mocking in tests. I added some comments there.

austinkelleher commented on src/lib/rules/html-checker/html-checker.ts:
```diff
@@ -0,0 +1,107 @@
+            try {
+                await Promise.all(reportPromises);
+            } catch (e) {
+                debug(`Error reporting the html checker results.`);
```
Same here. I think we should log the error too.

sarvaje commented on .sonarrc:
```diff
@@ -7,14 +7,15 @@
         }
     },
-    "formatter": "stylish",
+    "formatter": "codeframe",
```
I think the formatter by default is `stylish`

molant commented:
@qzhou1607 if you are using https://www.npmjs.com/package/html-validator it had at least an option to ignore in the configuration object. That property should be exposed somehow.

qzhou1607 commented:
@molant In `html-validator`, currently the setting of `ignore` is only supported if the output `format` is set to `text`. We have to use `json` here for further parsing. It was not difficult to implement `ignore` in our code though, so I added that. 

>@qzhou1607 if you are using https://www.npmjs.com/package/html-validator it had at least an option to ignore in the configuration object. That property should be exposed somehow.

I added documentation also.
>I don't see a page with the description of the rule. Remember to use the new-rule parameter you added to dogfood it and adjust if necessary!

I had issues with implementing async iterator. It looks like I need to convert iterables (array) into async iterables first using a generator, and then do the `for..await..of` loop. But for some reason, the rule can't be exported right whenever I drop the generator in the `create` function. Am I looking at the wrong direction?
>Can we add support for async iterator? They should be working on typescript 2.3 (check the release notes, there might be some hack to do to make it work everywhere).

qzhou1607 commented:
@molant @sarvaje @alrra I looked into `async iterator` more, still having trouble with async generators. The async generator function in typescript was directly copied when compiled and throws an error when running in Node. The `lib` setting I have in `tsconfig.json` is 
```
        "lib": [
            "es2017",
            "esnext",
            "esnext.asynciterable",
            "dom",
            "dom.iterable"
        ]
```
I also tried polyfill `Symbol.asyncIterator` according to [this post](https://stackoverflow.com/questions/43694281/ts2318-cannot-find-global-type-asynciterableiterator-async-generator) using `core-js`, still no luck, same error. So I'll keep the current code for now until I have more information about the async generators.

molant commented:
> I also tried polyfill `Symbol.asyncIterator` according to [this post](https://stackoverflow.com/questions/43694281/ts2318-cannot-find-global-type-asynciterableiterator-async-generator) using `core-js`, still no luck, same error. So I'll keep the current code for now until I have more information about the async generators.

Sounds good to me!

qzhou1607 commented:
@molant Yep working on it. If we give user this option we should probably ask for the output interface (`firstColumn`, `subType` and etc.) as well, otherwise the output can't be consumed by `context.report` and our formatter.

molant commented:
> ce (firstColumn, subType and etc.) as well, otherwise the output can't be consumed by context.report and our formatter.

Not sure I understand why we need that. We are only changing the end point to use.

qzhou1607 commented:
@molant A different service/endpoint may have a different REST interface (like using  `column` instead of `firstColumn` as the property name). This inconsistency could raise error when we process the output from different services. Or should we assume that other service endpoints always have the same output format as the default validator? If not, throws an error?
>Not sure I understand why we need that. We are only changing the end point to use.

molant commented:
We will be using the same code as the official one. There are some instructions on how to set it up.

qzhou1607 commented:
@molant validator option was added.

molant commented:
@sarvaje @AustinKelleher this seems to be ready for another review.

qzhou1607 commented on src/lib/rules/html-checker/html-checker.ts:
```diff
@@ -0,0 +1,155 @@
+            validator: ''
+        };
+
+        type Error = { // eslint-disable-line no-unused-vars
```
@sarvaje Done.

alrra commented on docs/user-guide/rules/html-checker.md:
````diff
@@ -0,0 +1,68 @@
+
+```json
+"html-checker": ["error", {
+    "ignore": ["Invalid attribute", "Invalid tag"]
````
We should use more specific example here. These are too general, and it might seem like one has to ignore types/groups of errors, and cannot ignore very specific ones.

alrra commented on docs/user-guide/rules/html-checker.md:
````diff
@@ -0,0 +1,68 @@
+}]
+```
+
+You can also override the default validator by passing in the endpoint of an alternative validator. However, you need to make sure that this alternative validator exposes the same REST interface as the default one.
````
`same REST interface` should link to the API.

alrra commented on docs/user-guide/rules/index.md:
```diff
@@ -16,6 +16,7 @@
 * [`content-type`](content-type.md)
 * [`highest-available-document-mode`](highest-available-document-mode.md)
 * [`no-friendly-error-pages`](no-friendly-error-pages.md)
+* [`html-checker`](html-checker.md)
```
Always use alphabetical order. Should be before `no-friendly-error-pages`.

alrra commented on src/lib/rules/html-checker/html-checker.ts:
```diff
@@ -0,0 +1,155 @@
+
+            scanOptions.validator = validator;
+
+            // Up to now, the `ignore` setting in `html-validator` only works if `format` is set to `text`
```
If this is a simple change, maybe we should offer and contribute back the fix to it?

alrra commented on src/lib/rules/html-checker/html-checker.ts:
```diff
@@ -0,0 +1,155 @@
+                }
+            }
+        }],
+        worksWithLocalFiles: false
```
It does not?

alrra commented on docs/user-guide/rules/html-checker.md:
```diff
@@ -0,0 +1,68 @@
+unintended mistakes which might have otherwise been missed.
+Adhering to the W3C' standards has a lot to offer to both the
+developers and the web users: It provides better browser compatibility,
+helps to avoid potential problems with accessibility/usability, and makes it easier for future maintainance.
```
Were possible, try to keep the text under 80 characters / line. The same goes for the lines below.

alrra commented on docs/user-guide/rules/html-checker.md:
````diff
@@ -0,0 +1,68 @@
+
+```json
+"html-checker": ["error", {
+    "validator": "http://html5.validator.nu"
````
`httpS`

alrra commented on src/lib/rules/html-checker/html-checker.ts:
```diff
@@ -0,0 +1,155 @@
+
+        const loadRuleConfig = () => {
+            const ignore = (context.ruleOptions && context.ruleOptions.ignore) || [];
+            const validator = (context.ruleOptions && context.ruleOptions.validator) || 'http://validator.w3.org/nu/';
```
`httpS`

alrra commented on src/lib/rules/html-checker/html-checker.ts:
```diff
@@ -0,0 +1,155 @@
+
+const rule: IRuleBuilder = {
+    create(context: RuleContext): IRule {
+        /** The promise that represents the scan by html checker. */
```
`html checker`, `Html Checker`, ... =>  Let's just use `HTML checker` consistently here, and throughout this file.

alrra commented on src/lib/rules/html-checker/html-checker.ts:
```diff
@@ -0,0 +1,155 @@
+    meta: {
+        docs: {
+            category: 'Interoperability',
+            description: 'Validating html using `the Nu html checker`'
```
> Validate HTML using 'the Nu html checker'

alrra commented on tests/lib/rules/html-checker/tests.ts:
```diff
@@ -0,0 +1,189 @@
+const ruleName = getRuleName(__dirname);
+const exampleUrl = 'https://example.com/';
+const validatorError = 'error';
+const defaultValidator = 'http://validator.w3.org/nu/';
```
`httpS`

alrra commented on tests/lib/rules/html-checker/tests.ts:
```diff
@@ -0,0 +1,189 @@
+const exampleUrl = 'https://example.com/';
+const validatorError = 'error';
+const defaultValidator = 'http://validator.w3.org/nu/';
+const configValidator = 'http://html5.validator.nu';
```
`httpS`

alrra commented on tests/lib/rules/html-checker/tests.ts:
```diff
@@ -0,0 +1,189 @@
+    const mockedChecker = (scanOptions) => {
+        let responseMessages;
+
+        if (response.pass) { // No errors/warnings are detected in the target html
```
`html` => Use `HTML` here, and everywhere else.

alrra commented:
Merged with some modifications in 9f73cc9b6071193a12933d756883738a0308a6b6.
